src/Lean/Meta/ExprDefEq.lean

```diff
@@ -1957,6 +1957,9 @@ private def isDefEqProj : Expr → Expr → MetaM Bool
 where
   /-- If `structName` is a structure with a single field and `(?m ...).1 =?= v`, then solve constraint as `?m ... =?= ⟨v⟩` -/
   isDefEqSingleton (structName : Name) (s : Expr) (v : Expr) : MetaM Bool := do
+    let some ctorVal := getStructureLikeCtor? (← getEnv) structName | return false
```

Comment:
I see two changes:
- The code move up. It was after the `if isClass ...`. Why?
- Replaced `getStructureCtor` with `getStructureLikeCtor?`. The motivation is clear to me: `getStructureCtor` now accepts recursive structures.

Reply:
> The code move up. It was after the if isClass .... Why?

I do not have a strong motivation, I will move it back.

Reply:
(Weak motivation: I found it easier to read with `isDefEqSingleton` checking that `structName` was a structure first.)